src/Microsoft.TemplateEngine.Cli/HelpForTemplateResolution.cs

```diff
@@ -0,0 +1,147 @@
+            {
+                Reporter.Error.WriteLine(string.Format(LocalizableStrings.NoTemplatesMatchName, templateName));
+                Reporter.Error.WriteLine();
+                return true;
+            }
+
+            // scp (2017-08-28): not certain anything in this loop can happen anymore, due to the way filtering occurs.
+            foreach (IReadOnlyList<ITemplateMatchInfo> templateGroup in contextProblemMatchGroups)
+            {
+                // all templates in a group should have the same context & name
+                if (templateGroup[0].Info.Tags != null && templateGroup[0].Info.Tags.TryGetValue("type", out ICacheTag typeTag))
+                {
+                    MatchInfo? matchInfo = WellKnownSearchFilters.ContextFilter(context)(templateGroup[0].Info);
+                    if ((matchInfo?.Kind ?? MatchKind.Mismatch) == MatchKind.Mismatch)
+                    {
+                        Reporter.Error.WriteLine(string.Format(LocalizableStrings.TemplateNotValidGivenTheSpecifiedFilter, templateGroup[0].Info.Name).Bold().Red());
+                        anythingReported = true;
+                    }
+                }
+                else
+                {   // this really shouldn't ever happen. But better to have a generic error than quietly ignore the partial match.
+                    Reporter.Error.WriteLine(string.Format(LocalizableStrings.GenericPlaceholderTemplateContextError, templateGroup[0].Info.Name).Bold().Red());
+                    anythingReported = true;
+                }
+            }
+
+            if (remainingPartialMatchGroups.Count > 0)
+            {
+                Reporter.Error.WriteLine(LocalizableStrings.TemplateMultiplePartialNameMatches.Bold().Red());
+                anythingReported = true;
+            }
+
+            Reporter.Error.WriteLine();
+            return anythingReported;
+        }
+
+        // Returns a list of the parameter names that are invalid for every template in the input group.
+        public static void GetParametersInvalidForTemplatesInList(IReadOnlyList<ITemplateMatchInfo> templateList, out IReadOnlyList<string> invalidForAllTemplates, out IReadOnlyList<string> invalidForSomeTemplates)
+        {
+            IDictionary<string, int> invalidCounts = new Dictionary<string, int>();
+
+            foreach (ITemplateMatchInfo template in templateList)
+            {
+                foreach (string paramName in template.GetInvalidParameterNames())
+                {
+                    if (!invalidCounts.ContainsKey(paramName))
+                    {
+                        invalidCounts[paramName] = 1;
+                    }
+                    else
+                    {
+                        invalidCounts[paramName]++;
+                    }
+                }
+            }
+
+            invalidForAllTemplates = invalidCounts.Where(x => x.Value == templateList.Count).Select(x => x.Key).ToList();
```

Comment:
I think there's an extension somewhere for doing bucketing - we're iterating the set twice for matching counts vs non-matching counts, which we could actually do in one pass